src/core/CoreModuleDirector.ts

```diff
@@ -160,25 +160,39 @@ export class CoreModuleDirector {
     return modelStores.pushSubscriptions.models as { [key: string]: OSModel<SupportedSubscription> };
   }
 
+  private async getPushSubscriptionModelByCurrentToken()
+    :Promise<OSModel<SupportedSubscription> | undefined> {
+      logMethodCall("CoreModuleDirector.getPushSubscriptionModelByCurrentToken");
+      const pushToken = await MainHelper.getCurrentPushToken();
+      if (pushToken) {
+        return this.getSubscriptionOfTypeWithToken(ModelName.PushSubscriptions, pushToken);
+      }
+      return undefined;
+  }
+
+  // Browser may return a different PushToken value, use the last-known value as a fallback.
+  //   - This happens if you disable notification permissions then re-enable them.
+  private async getPushSubscriptionModelByLastKnownToken()
+    :Promise<OSModel<SupportedSubscription> | undefined> {
+    logMethodCall("CoreModuleDirector.getPushSubscriptionModelByLastKnownToken");
+    const { lastKnownPushToken } = await Database.getAppState();
+    if (lastKnownPushToken) {
+      return this.getSubscriptionOfTypeWithToken(ModelName.PushSubscriptions, lastKnownPushToken);
+    }
+    return undefined;
+  }
+
   /**
    * Gets the current push subscription model for the current browser.
    * @returns The push subscription model for the current browser, or undefined if no push subscription exists.
    */
-  // TO DO: make synchronous by making getting the current push token synchronous
-  public async getCurrentPushSubscriptionModel(): Promise<OSModel<SupportedSubscription> | undefined> {
+  public async getPushSubscriptionModel()
```

Comment:
Did you think about ways to make getting the push token synchronous? This has the potential to make tons of dependent functions synchronous.

Reply:
The native browse API to get the current token `registration.pushManager.getSubscription()` returns a `Promise` so it isn't possible.
* If cached it and use a cache value that could work around that, but then it won't be current and it has a risk of being out of date then.